Moving content from composer to use-cases (#1817)
saucelabs/sauce-docs: docs/api-testing/use-cases/integration-test.md

---
id: integration-test
title: Build an Integration Test
sidebar_label: Build an Integration Test
description: How to generate an integration test
---

import useBaseUrl from '@docusaurus/useBaseUrl';

One of the core features of the platform is the ability to create proper Integration Tests. An Integration Test is a test in which you examine a complete user flow, simulating what an API consumer would experience. Integration testing is critical for creating a strong API testing strategy. While only testing individual endpoints is a good start, this method will miss a large number of problems that occur when all services need to work together.

## What You'll Need

- A Sauce Labs account ([Log in](https://accounts.saucelabs.com/am/XUI/#login/) or sign up for a [free trial license](https://saucelabs.com/sign-up)).
- An existing API Testing Project. For details on how to create one, see [API Testing Quickstart](/api-testing/quickstart/).

## Test Interactions between Endpoints

> For the purposes of this guide, will keep our example simple and work only with two API endpoints but your integration test can be longer.

Reviewer: ```suggestion
For the purposes of this guide, the example is simple and works with two API endpoints, but your integration test can be longer.
```

> In the following example, we'll use an API endpoint that produces an array of all the available products (http://demoapi.apifortress.com/api/retail/product) and another endpoint that shows the details of a specific product based on its ID (http://demoapi.apifortress.com/api/retail/product/${id}).

Reviewer: ```suggestion
In the following example, uses an API endpoint that produces an array of all the available products (http://demoapi.apifortress.com/api/retail/product) and another endpoint that shows the details of a specific product based on its ID (http://demoapi.apifortress.com/api/retail/product/${id}).
```

2. (Optionally) Add an assert-is component to verify that `productsPayload` is an array.

:::tip
You can use the Comment component and print the whole payload response inside your test report by writing ${productsPayload}
:::

3. Add an `each` component and reference the `productsPayload` object.

:::note
In a scenario in which the response contains many products, it may be useful to pick a few at random by using [pick(n)](/api-testing/composer/expressions/#pick).
:::

4. Test the response payload for the endpoint by adding assertions.

5. Add a new `Set(variable)` assertion to set the `id` variable as every single product that is returned:
   - Variable (the variable name) - for example `id`
   - Mode (the variable type) - for example `String`
   - Value (the value the variable will contain) - for example `${_1.id}`

The system uses `_1` automatically when recognizing a subroutine, which makes it easier when there are multiple sub-levels.

> Below how your test might look like till the above step:

Reviewer: Not sure what you're trying to say here. I tried fix it. 
```suggestion
This is how your test might look after you set the `id` variable:
```
